course5/homework10/linked_list.h

```diff
@@ -0,0 +1,221 @@
+        }
+        mSize++;
+    }
+
+    template< class... Args >
+    void emplace_front(Args&&... args)
+    {
+        if (mSize > 0)
+        {
+            LinkedList<T> new_node;
+            new_node.data = std::move(this->data);
+            new_node.next = std::move(this->next);
+            data = std::make_unique<T>(args...);
+            next = std::make_unique<LinkedList<T>>(std::move(new_node));
+        }
+        else
+        {
+            data = std::make_unique<T>(args...);
+        }
+        mSize++;
+    }
+
+    void pop_front()
+    {
+        if (next)
+        {
+            LinkedList<T> new_head = std::move(*next.get());
+            data = std::move(new_head.data);
+            next = std::move(new_head.next);
+            mSize--;
+        }
+    }
+
+    void print() const
+    {
+        std::cout << "------------------------------------------------------------------------------" << std::endl;
+        std::cout << "getting this" << std::endl;
+        LinkedList<T> current = *this;
+        std::cout << "got this" << std::endl;
+        for (auto i = 0; i < mSize; i++)
+        {
+            std::cout << i << " [" << current.front() << "]" << std::endl;
+            if (current.next)
+            {
+                std::cout << "getting next ptr" << std::endl;
+                LinkedList<T> *newcurrent = current.next.get();
+                std::cout << "getting next current" << std::endl;
+                current = *(newcurrent);
+                std::cout << "got next" << std::endl;
+            }
+            else
+            {
+                std::cout << "breaking" << std::endl;
+                break;
+            }
+        }
+        std::cout << std::endl;
+    }
+
+    iterator begin() { return iterator(*this); }
+    iterator end() { return iterator(nullptr); }
+
+private:
+    std::unique_ptr<T> data;
+    std::unique_ptr<LinkedList<T>> next;
+    int mSize = 0;
+
+}; // LinkedList
+
+template<class T>
+class LinkedListIter {
+    public:
+        LinkedListIter(LinkedList<T>& list_arg) : mList(list_arg) {}
+        bool operator==(LinkedListIter & cmp)
+        {
+            return (mList == cmp.mList);
+        }
+        bool operator!=(LinkedListIter & cmp)
+        {
+            return !(this == cmp);
+        }
+        T & operator*() {
+            // if (my_list.data)
+                return *mList.data.get();
+            // else
+                // return *nullptr;
+        }
+        LinkedListIter & operator++() {
+            // if (mList.next)
+            // {
+                mList = *mList.next.get();
+                return *this;
+            // }
+            // else
+            //     return nullptr;
+        }
+    private:
+        LinkedList<T>& mList;
+};
+
+// template<class T>
+// bool operator==(const LinkedList<T>& cmp1, const LinkedList<T>& cmp2)
+// {
+//     std::cout << " binary " << std::endl;
+//     return cmp1.operator==(&cmp2);
```

Comment:
Back to my `==` issues, I was thinking I could also define a public `check_next__nodes_are_equal` method, but I don't really want any idea of the `next` to be public. I'd rather force the compiler to call the unary version.